src/ApplicationInsights.Kubernetes/Extensions/KubernetesServiceCollectionBuilder.cs

```diff
@@ -2,142 +2,140 @@
-            // For backward compatibility, $APPINSIGHTS_KUBERNETES_POD_NAME has been agreed upon to allow customize pod name with downward API.
-            serviceCollection.TryAddEnumerable(ServiceDescriptor.Scoped<IPodNameProvider, UserSetPodNameProvider>());
-            // $Hostname will be overwritten by Kubernetes to reveal pod name: https://kubernetes.io/docs/concepts/containers/container-environment/#container-information.
-            serviceCollection.TryAddEnumerable(ServiceDescriptor.Scoped<IPodNameProvider, HostNamePodNameProvider>());
+            // For example, EnvironmentVariableContainerIdProvider will take precedence, then CGroupContainerIdProvider on Linux.
+            serviceCollection.TryAddEnumerable(ServiceDescriptor.Scoped<IContainerIdProvider, CGroupContainerIdProvider>());                 // Then CGroupV1
+            serviceCollection.TryAddEnumerable(ServiceDescriptor.Scoped<IContainerIdProvider, ContainerDMountInfoContainerIdProvider>());    // Then ContainerD
+            serviceCollection.TryAddEnumerable(ServiceDescriptor.Scoped<IContainerIdProvider, DockerEngineMountInfoContainerIdProvider>());  // Then DockerEngine
         }
-
-        /// <summary>
-        /// Registers K8s environment factory.
-        /// </summary>
-        protected virtual void RegisterK8sEnvironmentFactory(IServiceCollection serviceCollection)
+        else if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+        {
+            // Notes: pay attention to the order. Injecting uses the order of registering in this case.
+            // For example, EnvironmentVariableContainerIdProvider will take precedence, then EmptyContainerIdProvider on Windows.
+            serviceCollection.TryAddEnumerable(ServiceDescriptor.Scoped<IContainerIdProvider, EmptyContainerIdProvider>());
+        }
+        else
         {
-            serviceCollection.TryAddScoped<IK8sEnvironmentFactory, K8sEnvironmentFactory>();
-            serviceCollection.TryAddSingleton<IK8sEnvironmentHolder>(_ => K8sEnvironmentHolder.Instance);
-            serviceCollection.AddHostedService<K8sInfoBackgroundService>();
+            _logger.LogError("Unsupported OS.");
         }
+
+        serviceCollection.TryAddScoped<IContainerIdNormalizer, ContainerIdNormalizer>();
+
+        // Notes: pay attention to the order. Injecting uses the order of registering in this case.
+        // For backward compatibility, $APPINSIGHTS_KUBERNETES_POD_NAME has been agreed upon to allow customize pod name with downward API.
+        serviceCollection.TryAddEnumerable(ServiceDescriptor.Scoped<IPodNameProvider, UserSetPodNameProvider>());
+        // $Hostname will be overwritten by Kubernetes to reveal pod name: https://kubernetes.io/docs/concepts/containers/container-environment/#container-information.
+        serviceCollection.TryAddEnumerable(ServiceDescriptor.Scoped<IPodNameProvider, HostNamePodNameProvider>());
```

Comment:
Not relevant to this PR specifically but I'm curious about why you're adding a provider class for each possible environment variable. It seems like a lot of added ceremony unless those managers are expected to get a lot more complex. 

I wonder if you could do the same thing with a readonly queue in the `EnvironmentVariablePodNameProviderBase.cs` class and remove the individual implementations.

Reply:
That is a very good question.

My rationale behind it is the OC principal. By code history, UserSetPodName was implemented way earlier than the HostName. When the time hostname was implemented later, I had two choices:
1. Change the existing implementation of the user set environment variable class and update code there;
2. Extend the implementation by adding another class.

In this case, I actually am okay with either choice. Nothing complex at all and I don't expect any bugs.
For more complex scenarios, though, I would prefer a new implementation to change the existing code.

Does that answer the question? Please let me know and we can have a further discuss on it.

Reply:
Totally makes sense. Probably not worth spending time on unless it comes up more. Thanks for the answer.